src/HealthGPS.Console/csvparser.cpp

```diff
@@ -79,101 +81,91 @@ hc::DoubleDataTableColumnBuilder parse_double_column(std::string name,
+
+        auto is_match = [&col_name](const auto &csv_col_name) {
+            return hc::case_insensitive::equals(col_name, csv_col_name);
         };
 
-        if (auto it = std::find_if(headers.begin(), headers.end(), is_match); it != headers.end()) {
-            csv_cols[pair.first] = *it;
+        auto it = std::find_if(headers.begin(), headers.end(), is_match);
+        if (it != headers.end()) {
+            csv_column_map[col_name] = *it;
         } else {
-            mismatch++;
-            fmt::print(fg(fmt::color::dark_salmon), "Column: {} not found in dataset.\n",
-                       pair.first);
+            success = false;
+            fmt::print(fg(fmt::color::dark_salmon), "Column: {} not found in dataset.\n", col_name);
         }
     }
 
-    if (mismatch > 0) {
-        return false;
+    if (!success) {
+        throw std::runtime_error("Required columns not found in dataset.");
     }
 
-    auto result = true;
-    for (auto &pair : csv_cols) {
-        auto col_type = hc::to_lower(columns.at(pair.first));
+    for (const auto &[col_name, csv_col_name] : csv_column_map) {
+        std::string col_type = hc::to_lower(file_info.columns.at(col_name));
 
         // Get raw data
-        auto data = doc.GetColumn<std::string>(pair.second);
+        auto data = doc.GetColumn<std::string>(csv_col_name);
+
         try {
             if (col_type == "integer") {
-                out_table.add(detail::parse_int_column(pair.first, data).build());
+                out_table.add(parse_int_column(col_name, data).build());
             } else if (col_type == "double") {
-                out_table.add(detail::parse_double_column(pair.first, data).build());
+                out_table.add(parse_double_column(col_name, data).build());
             } else if (col_type == "float") {
-                out_table.add(detail::parse_float_column(pair.first, data).build());
+                out_table.add(parse_float_column(col_name, data).build());
             } else if (col_type == "string") {
-                out_table.add(detail::parse_string_column(pair.first, data).build());
+                out_table.add(parse_string_column(col_name, data).build());
             } else {
                 fmt::print(fg(fmt::color::dark_salmon), "Unknown data type: {} in column: {}\n",
-                           col_type, pair.first);
-                result = false;
+                           col_type, col_name);
+                success = false;
             }
         } catch (const std::exception &ex) {
-            fmt::print(fg(fmt::color::red), "Error passing column: {}, cause: {}\n", pair.first,
+            fmt::print(fg(fmt::color::red), "Error passing column: {}, cause: {}\n", col_name,
                        ex.what());
-            result = false;
+            success = false;
         }
     }
 
-    return result;
+    if (!success) {
+        throw std::runtime_error("Error parsing dataset.");
+    }
+
+    return success;
```

Comment:
I should really just return `out_table` instead of the `success` bool, now that I have exceptions. Doing so is not entirely trivial, though, as the default move constructor for the table cannot be constructed, possibly due to the mutex. It's possible we can just manually define the move constructor, which will `std::move(mutex)`. Not sure. Needs trying.

Reply:
Are there any consequences of wrapping mutexes in unique pointers? Besides still not having copy ctors?

Reply:
Not really. You get one extra heap allocation of a few bytes when you create the object, but I don't think we have to worry about that too much :stuck_out_tongue: